### PR TITLE
Fix s

### PR DESCRIPTION
**Pull request by ulrikeyang**
This pull request fixes the hopscotch problem. 
The generation of the strength matrix for nodal coarsening combined with unknown-based approach has been rewritten.
A new option HYPRE_BoomerAMGSetKeepSameSign has been added, which allows different generation of this strength matrix.
The default is 0, i.e. turned off, since this is necessary for good convergence for extended+i interpolation, but it hurts
classical interpolation and MM-extended interpolation (the latter needs further investigation). In the future I would like to get rid of this option, but for right now it might be useful for certain systems problems.
I have also included the unknown-based approach into MM-extended, MM-extended+i , MM-extended+e, and 2-stage MM-extended+e interpolation. My inclusion into 2-stage MM-extended interpolation doesn't work well, so I took it out (also needs further investigation).
I removed the use of col_offd_S_to_A and the functionality of HYPRE_BoomerAMGSetSCommPkgSwitch. The function is still there, to not break users' code in case somebody actually uses it, but it doesn't do anything.

### CODE REVIEW

File: src/test/TEST_ij/agg_interp.saved [hunk 1]
```diff
@@ -74,7 +74,7 @@ Final Relative Residual Norm = 5.870410e-09
 
 
 Iterations = 16
-Final Relative Residual Norm = 5.087936e-09
+Final Relative Residual Norm = 5.163837e-09
```
Comment (rfalgout) [inline src/test/TEST_ij/agg_interp.saved:77]: Is it clear why these residuals are different now?  Just being cautious.

**Review by rfalgout — APPROVED** [review]
Hi @ulrikeyang .  Nice work on all these changes!  It looks good to me, though I did not look through the code logic at all, since others are more qualified for that.  The only question I have is on the regression testing and the differences in saved file numbers.  It's always good to be extra cautious when doing this, because even small changes can sometimes indicate underlying issues.  If you are confident about why they differ, then that's good enough for me.

Thanks again!

File: src/test/TEST_ij/agg_interp.saved (hunk 1 above)
Comment (ulrikeyang) [reply src/test/TEST_ij/agg_interp.saved:77]: Yes, this was a systems problem, as are the ones below, so I would have expected some change based on the changes I made.

Comment (rfalgout) [reply src/test/TEST_ij/agg_interp.saved:77]: Perfect!

File: src/parcsr_ls/par_2s_interp.c [hunk 2]
```diff
@@ -281,27 +314,83 @@ hypre_BoomerAMGBuildModPartialExtInterpHost( hypre_ParCSRMatrix  *A,
       {
          if (CF_marker[i] == -2)
          {
-            for (j=A_diag_i[i]; j < A_diag_i[i+1]; j++)
+            /*if (num_functions > 1)
```
Comment (liruipeng) [inline src/parcsr_ls/par_2s_interp.c:317]: This is commented out but not in the next one ext+e, which is fine. Just want to make sure it is what you want. Thanks!

File: src/parcsr_ls/par_nodal_systems.c [hunk 3]
```diff
@@ -798,74 +798,64 @@ hypre_BoomerAMGCreateNodalA(hypre_ParCSRMatrix    *A,
...
-   HYPRE_Int           SN_num_nonzeros_offd;
+   HYPRE_Int           num_cols_offd_A = hypre_CSRMatrixNumCols(A_offd);
+   HYPRE_Int           A_num_nonzeros_diag;
+   HYPRE_Int           A_num_nonzeros_offd;
    HYPRE_Int           S_num_nonzeros_diag;
    HYPRE_Int           S_num_nonzeros_offd;
    HYPRE_BigInt        global_num_vars;
    HYPRE_BigInt        global_num_cols;
    HYPRE_BigInt        global_num_nodes;
+   HYPRE_Int           nnz, S_cnt, in;
```
Comment (liruipeng) [inline src/parcsr_ls/par_nodal_systems.c:858]: This is not a new issue in this PR. Line 871 `if (nodal < 0) {}`: 1. not sure what's the meaning of `nodal < 0`, probably not documented. 2. `dof_func` is allocated as size of `num_coarse_nodes*num_functions` only for C points, which does not match the comment above this function, "scalar version of dof_array"

**Review by liruipeng — APPROVED** [review]
Looks good to me. Thanks!

File: src/parcsr_ls/aux_interp.c [hunk 4]
```diff
@@ -354,7 +354,15 @@ static HYPRE_Int hypre_new_offd_nodes(HYPRE_BigInt **found, HYPRE_Int num_cols_A
         {
           if (!hypre_UnorderedBigIntSetContains(&set, big_i1))
           {
-            Sop_j[j] = -hypre_UnorderedBigIntMapGet(&col_map_offd_inverse, big_i1) - 1;
+            HYPRE_Int k = hypre_UnorderedBigIntMapGet(&col_map_offd_inverse, big_i1);
+            if (-1 == k)
+            {
+               hypre_UnorderedBigIntSetPut(&set, big_i1);
```
Comment (liruipeng) [inline src/parcsr_ls/aux_interp.c:360]: Is it still possible to happen after your changes that make sure S is a subset of A? Just curious. Thanks!

**Comment by ulrikeyang** [comment]
It should not happen unless we set HYPRE_BoomerAMGSetKeepSameSign to 1. In that case, we could still have a problem, but it should not crash, it would be a potential convergence issue, but convergence is tricky in general for systems problems. Sometimes it might even help.
Note that hopscotch will now act exactly like the non-hopscotch version for this case.

**Comment by ulrikeyang** [comment]
Yes , I commented this out on purpose, since the systems version seem to make things worse for this case on all the regression problems. It’s something that needs to be investigated, but I didn’t have time to do this, so I thought it’s better to leave it out for now.

**Review by oseikuffuor1 — APPROVED** [review]
Looks good to me. Thanks @ulrikeyang 
.

**Pull request merged**